Feat/configurations guide (#776)
tauri-apps/tauri-docs: docs/guides/architecture/configuration-files.md

Another thing to take note of is the `features=[]` portion of the `tauri` dependency. Running `tauri dev` and `tauri build` will automatically manage which features need to be enabled in your project based on the `"allowlist"` properties you set in `tauri.conf.json`.

When you build your application a `Cargo.lock` file is produced. This file is used primarily for ensuring that the same dependencies are used later on as were used during development. Since you are developing a Tauri app, this file should be committed into your source repository. Only Rust libraries should omit comitting this file.

To learn more about `Cargo.toml` you can read more in the [official documentation][cargo-manifest].

## `package.json`

This is the Node.js package file. If the frontend of a Tauri app is developed using Node.js-based technologies this file is used to configure the frontend such as which dependencies are used. It's also common to use the `"scripts"` section to store the command used to launch the frontend used by your Tauri application.

An example of a barebones `package.json` file for a Tauri project might look a little something like this:

```json title=package.json
{
  "scripts": {

>     "dev": "command-for-your-framework"

Reviewer: ```suggestion
    "dev": "command-for-your-framework",
    "tauri": "tauri"
```

---
sidebar_position: 1
---

# Configuration Files

Since Tauri is a toolkit for building applications there can be many files to configure project settings. Some common files that you may run across are `tauri.conf.json`, `package.json` and `Cargo.toml`. We briefly explain each on this page to help point you in the right direction for which file to modify.

> ## `tauri.conf.json`

Reviewer 2: ```suggestion
## `tauri.conf.json`, `tauri.conf.json5` and `Tauri.toml`
```

> ## `tauri.conf.json`

Author: ```suggestion
## Tauri Config (`tauri.conf.json`, `tauri.conf.json5`, or `Tauri.toml`)
```

> This is the file used by the Tauri process. You can define build settings (such as the [command run prior to `tauri build`][before-build-command] or [`tauri dev`][before-dev-command]), set the [name and version of your app][package-config], [control the Tauri process][tauri-config], and configure any plugin settings. You can find all of the options in the [`tauri.conf.json` API reference].

Author: ```suggestion
This is the file used by the Tauri process. You can define build settings (such as the [command run prior to `tauri build`][before-build-command] or [`tauri dev`][before-dev-command]), set the [name and version of your app][package-config], [control the Tauri process][tauri-config], and configure any plugin settings. You can find all of the options in the [`tauri.conf.json` API reference].

:::note

The default Tauri config format is `.json`. `.json5` and `.toml` formats can be enabled by adding the feature flag `config-json5` or `config-toml` (respectively) to the `tauri` dependency in `Cargo.toml`. The structure and values are the same across all formats, however the formatting should be consistent with the respective file format.

`.toml` is only available from Tauri 1.1.0 and above.

:::
```

Reviewer: This should probably mention that json5 and toml have their own feature flags _which are disabled by default_. The flags are `config-json5` and `config-toml`

Author: Where do those flags go? In the `Cargo.toml` file?

Reviewer: Yeah on the tauri dependency, like this thingy here https://github.com/tauri-apps/tauri/blob/dev/examples/sidecar/src-tauri/Cargo.toml#L14

Author: And added to the regular dependency, not the build dependency, right?

Reviewer: yeah

Reviewer: Actually tauri-build has that flag too, but i've never seen it being used 🤔 Gonna double check if it's enough on tauri

## `Cargo.toml`

> Cargo's manifest file used to declare Rust crates your app depends on, metadata about your app, and other Rust-related features. If you do not intend to do backend development using Rust for your app then you may not modifying it much, but it's important to know that it exists and what it does.

Reviewer: ```suggestion
Cargo's manifest file is used to declare Rust crates your app depends on, metadata about your app, and other Rust-related features. If you do not intend to do backend development using Rust for your app then you may not be modifying it much, but it's important to know that it exists and what it does.
```

Below is an example of a barebones `Cargo.toml` file for a Tauri project:

```toml title=Cargo.toml
[package]
name = "app"
version = "0.1.0"
description = "A Tauri App"
authors = ["you"]
license = ""
repository = ""
default-run = "app"
edition = "2021"
rust-version = "1.57"

[build-dependencies]
tauri-build = { version = "1.0.0" }

[dependencies]
serde_json = "1.0"
serde = { version = "1.0", features = ["derive"] }
tauri = { version = "1.0.0", features = [ "api-all" ] }

[features]
# by default Tauri runs in production mode
# when `tauri dev` runs it is executed with `cargo run --no-default-features` if `devPath` is an URL
default = [ "custom-protocol" ]
# this feature is used used for production builds where `devPath` points to the filesystem
# DO NOT remove this
custom-protocol = [ "tauri/custom-protocol" ]
```

> The most important parts to take note of are the `tauri-build` and `tauri` dependencies. Their versions must be the same as the Tauri CLI in order to ensure that there won't be any issues running your app. If any of these are not the same version number you are highly likely to encounter issues.

Reviewer: welp, this is not correct anymore. And not that big of a problem now that the cli and tauri-build are more decoupled. There is still a requirement to have them somehow match, but no idea how one would phrase that.

Author: I really don't like that we have different version numbers in that monorepo, it makes things a lot harder to keep on top of. Let me take a stab at rephrasing this.

> The most important parts to take note of are the `tauri-build` and `tauri` dependencies. Their versions must be the same as the Tauri CLI in order to ensure that there won't be any issues running your app. If any of these are not the same version number you are highly likely to encounter issues.

Author: @FabianLars thoughts on this phrasing?
```suggestion
The most important parts to take note of are the `tauri-build` and `tauri` dependencies. Generally they must both be on the latest minor versions as the Tauri CLI, but this is not strictly required. If you encounter issues while trying to run your app you should check that any Tauri versions (`tauri` and `tauri-cli`) are on the latest versions for their respective minor releases.
```

Reviewer: Why did i not get a notification 😢 But yeah sounds good x)

Cargo version numbers use [Semantic Versioning]. Running `cargo update` will pull the latest available Semver-compatible versions of all dependencies. For example, if you specify `1.0.0` as the version for `tauri-build`, Cargo will actually detect and download version `1.0.4` because it is the latest Semver-compatible version available. Tauri will update the major version number whenever a breaking change is introduced, meaning you should always be capable of safely upgrading to the latest minor and patch versions without fear of your code breaking.

> If you want to use specific crate version you can use exact versions instead by prepending `=` to the version number of the dependency:

Reviewer: ```suggestion
If you want to use a specific crate version you can use exact versions instead by prepending `=` to the version number of the dependency:
```

To learn more about `Cargo.toml` you can read more in the [official documentation][cargo-manifest].

## `package.json`

This is the package file used by Node.js. If the frontend of a Tauri app is developed using Node.js-based technologies (such as `npm`, `yarn`, or `pnpm`) this file is used to configure the frontend dependencies and scripts.

An example of a barebones `package.json` file for a Tauri project might look a little something like this:

```json title=package.json
{
  "scripts": {
    "dev": "command-for-your-framework",
    "tauri": "tauri"
  },
  "dependencies": {
    "@tauri-apps/api": "^1.0",
    "@tauri-apps/cli": "^1.0"
  }
}
```

It's common to use the `"scripts"` section to store the command used to launch the frontend used by your Tauri application. The above file specifies the `dev` command that you can run using `yarn dev` or `npm run dev` to start the frontend framework.

> The dependencies object specifies which dependencies Node.js should downloaded when you run either `yarn` or `npm install` (in this case the Tauri CLI and API).

Reviewer: ```suggestion
The dependencies object specifies which dependencies Node.js should download when you run either `yarn` or `npm install` (in this case the Tauri CLI and API).
```
